lib/SILGen/SILGenLValue.cpp

```diff
@@ -2330,21 +2331,31 @@ namespace {
                  TypeKind == KPTK_WritableKeyPath ||
                  TypeKind == KPTK_ReferenceWritableKeyPath) {
         projectFn = SGF.getASTContext().getGetAtKeyPath();
+        sig = projectFn->getGenericSignature();
+        auto genericParams = sig.getGenericParams();
 
         auto keyPathTy = keyPathValue.getType().castTo<BoundGenericType>();
         assert(keyPathTy->getGenericArgs().size() == 2);
         assert(keyPathTy->getGenericArgs()[0]->getCanonicalType() ==
                BaseFormalType->getCanonicalType());
-        typeArgs.push_back(keyPathTy->getGenericArgs()[1]);
+        assert(genericParams.size() == 2);
+        auto secondGenParam = genericParams[1]->getCanonicalType()
+                                              ->castTo<SubstitutableType>();
+        map[secondGenParam] = keyPathTy->getGenericArgs()[1];
 
         keyPathValue = emitUpcastToKeyPath(SGF, loc, TypeKind, keyPathValue);
       } else {
         llvm_unreachable("bad key path kind for this component");
       }
 
-      auto subs = SubstitutionMap::get(projectFn->getGenericSignature(),
-                                       ArrayRef<Type>(typeArgs),
-                                       ArrayRef<ProtocolConformanceRef>());
+      sig = projectFn->getGenericSignature();
+      auto firstGenParam = sig.getGenericParams()[0]->getCanonicalType()
+                                                  ->castTo<SubstitutableType>();
+      map[firstGenParam] = BaseFormalType;
+
+      auto subs = SubstitutionMap::get(sig,
```

Comment:
It might be easier to use the get() that just takes arrays of replacement types and conformances

Reply:
In terms of lines-of-code, this version building a TypeSubstitutionMap ended up being more concise / clean.